### PR TITLE
feat(Dockerfile): use .nvmrc to build the image

### PR DESCRIPTION
**Pull request by PixnBits**
## Description

use .nvmrc to build the image
and run workflows, creating a single point to define the Node.js version to use and thus making upgrades easier

## Motivation and Context

Avoid the myriad of locations where Node.js versions are recorded, i.e. Dockerfile and workflows. package.json engines will still need manual updates though (perhaps that can be a workflow, or a unit test to ensure it is satisfied by the contents of `.nvmrc`?)

## How Has This Been Tested?

built locally, though not run, but did see the expected label used when manually running `$ docker build . --build-arg VERSION=18.17.0` and `docker build . --build-arg VERSION=$(cat .nvmrc)`

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (adding or updating documentation)
- [ ] Dependency update
- [ ] Security update

## Checklist:

- [ ] My change requires a change to the documentation and I have updated the documentation accordingly.
- [x] These changes should be applied to a maintenance branch.
- [ ] This change requires cross browser checks.
- [ ] Performance tests should be ran against the server prior to merging.
- [ ] This change impacts caching for client browsers.
- [ ] This change impacts HTTP headers.
- [ ] This change adds additional environment variable requirements for One App users.
- [ ] I have added the Apache 2.0 license header to any new files created.

## What is the Impact to Developers Using One App?

faster Node.js version updates? this is more for the maintainers so no impact to devs is the basic requirement 😅

### CODE REVIEW

File: __tests__/package.spec.js [hunk 1]
```diff
@@ -13,16 +13,69 @@
...
+        expect.assertions(1);
+        const rawNVM = await fs.readFile(path.join(PACKAGE_DIR_PATH, '.nvmrc'), 'utf8');
+        const { engines } = JSON.parse(await fs.readFile(path.join(PACKAGE_DIR_PATH, 'package.json')));
+        expect(satisfies(rawNVM, engines.node)).toBe(true);
+      });
+      it('is inline with .nvmrc', async () => {
+        expect.assertions(1);
+        const rawNVM = await fs.readFile(path.join(PACKAGE_DIR_PATH, '.nvmrc'), 'utf8');
+        const { engines } = JSON.parse(await fs.readFile(path.join(PACKAGE_DIR_PATH, 'package.json')));
+        expect(engines.node).toBe(`^${parse(rawNVM).major}`);
```
Comment (PixnBits) [inline __tests__/package.spec.js:77]: proof of work: https://github.com/americanexpress/one-app/actions/runs/5869749612/job/15915395595

do we want to keep the original `>=` or is `^` better? (my thought is with this check we'll be updating the field alongside the `.nvmrc` file, so shouldn't be an impediment to intentionally using a different major version of Node.js)

**Review by code-forger — APPROVED** [review]

**Review by code-forger — APPROVED** [review]

**Review by code-forger — CHANGES_REQUESTED** [review]
See comment

**Comment by code-forger** [comment]
This PR will look like its not building, because our settings require workflows that this PR removes.

Once this PR is ready and approved, the setting should be changed immediately before merge

File: .github/workflows/on-pr-and-trunk_one-app-unit-and-lint-tests.yml [hunk 2]
```diff
@@ -9,9 +9,6 @@ jobs:
   one-app-unit-tests:
     name: One App Unit and Lint Tests
     runs-on: ubuntu-latest
-    strategy:
-      matrix:
-        node-version: [16.x, "^18 < 18.16"]
```
[the remark this answers is not in the document]
Comment (dogpatch626) [reply .github/workflows/on-pr-and-trunk_one-app-unit-and-lint-tests.yml:14]: +1

File: .github/workflows/on-pr-and-trunk_one-app-unit-and-lint-tests.yml [hunk 3]
```diff
@@ -22,7 +19,7 @@ jobs:
       - name: Setup node
         uses: actions/setup-node@v3
         with:
-          node-version: ${{ matrix.node-version }}
+          node-version-file: .nvmrc
```
Comment (JAdshead) [inline .github/workflows/on-pr-and-trunk_one-app-unit-and-lint-tests.yml:22]: lets revert this change and above, apart from that LGTM!!
```suggestion
          node-version: ${{ matrix.node-version }}
```

File: .github/workflows/on-pr-and-trunk_one-app-unit-and-lint-tests.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (PixnBits) [reply .github/workflows/on-pr-and-trunk_one-app-unit-and-lint-tests.yml:14]: I was hoping to find a way to build that matrix with the .nvmrc file version, but that can be a different PR (if I do figure that out 😅)

[the remark this answers is not in the document]
Comment (PixnBits) [reply .github/workflows/on-pr-and-trunk_one-app-unit-and-lint-tests.yml:14]: changed in 527638f

File: .github/workflows/on-pr-and-trunk_one-app-unit-and-lint-tests.yml (hunk 3 above)
Comment (PixnBits) [reply .github/workflows/on-pr-and-trunk_one-app-unit-and-lint-tests.yml]: reverted in 527638f 🙇

**Review by JAdshead — DISMISSED** [review]

File: .nvmrc [hunk 4]
```diff
@@ -0,0 +1 @@
+18.17.0
```
Comment (PixnBits) [inline .nvmrc:1]: after #1087 this should be
```suggestion
18.17.1
```

Comment (PixnBits) [reply .nvmrc]: updated in 9e5b850

**Review by JAdshead — DISMISSED** [review]

File: .github/workflows/on-pr-and-trunk_one-app-unit-and-lint-tests.yml [hunk 5]
```diff
@@ -11,7 +11,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        node-version: [16.x, "^18 < 18.16"]
+        node-version: [16.x, 18.x]
```
Comment (PixnBits) [inline .github/workflows/on-pr-and-trunk_one-app-unit-and-lint-tests.yml:14]: from https://github.com/americanexpress/one-app/actions/runs/6042820335/job/16398714236?pr=1090
perhaps this should instead be
```suggestion
        node-version: [18.x]
```
until 20.x is LTS?

Comment (JAdshead) [reply .github/workflows/on-pr-and-trunk_one-app-unit-and-lint-tests.yml:14]: maybe revert the engines field. This still supports 16 for another week or so

Comment (PixnBits) [reply .github/workflows/on-pr-and-trunk_one-app-unit-and-lint-tests.yml:14]: whoops, forgot that change was me, good catch!

Comment (PixnBits) [reply .github/workflows/on-pr-and-trunk_one-app-unit-and-lint-tests.yml:14]: reverted the engines and tests of it in c63a856

**Review by JAdshead — DISMISSED** [review]

**Review by dogpatch626 — DISMISSED** [review]

File: Dockerfile [hunk 6]
```diff
@@ -1,8 +1,9 @@
+ARG VERSION=latest
```
Comment (Matthew-Mallimo) [inline Dockerfile:1]: will this not bring in node 20?

Comment (PixnBits) [reply Dockerfile:1]: It will bring in non-LTS versions too, if the build argument is not provided (see `docs/guides/Running-In-Production.md`). Many of the changes in this PR are to supply that argument 👍

Comment (PixnBits) [reply Dockerfile:1]: We could use a specific version as the default, instead of `latest` but that could be indirection and cause surprises? 🤔

Comment (PixnBits) [reply Dockerfile:1]: From https://hub.docker.com/_/node there is an `lts` tag, we could use that as the default?
```suggestion
ARG VERSION=lts
```

**Review by JAdshead — APPROVED** [review]

**Review by dogpatch626 — APPROVED** [review]
